docs/developer/container-goals.md

```diff
@@ -0,0 +1,297 @@
+
+According to the documentation, kaniko is meant to be run as an image. You can learn more about the rest of the arguments and optional flags from [the kaniko documentation](https://github.com/GoogleContainerTools/kaniko#additional-flags). In this example, we're:
+
+- passing the location of our Dockerfile (relative to the directory of the repository) (`--dockerfile`)
+- providing the path to directory that contains the Dockerfile ([`--context`](https://github.com/GoogleContainerTools/kaniko#kaniko-build-contexts))
+- indicating that we want to build the image _without_ pushing to a registry (`--no-push`)
+- taking a snapshot at the end of the build (`--single-snapshot`)
+- setting a destination directory for the resulting image (`--destination`)
+
+And finally, just as we only want to build our JAR in the Maven repositories our SDM is listening to, we only want to build our Docker image in locations with a `Dockerfile`. We will once again configure a push test, with one additional step: we'll make sure to run this goal _after_ we build the JAR file. You'll notice that we added a new step, `dependsOn`, which means that this goal executes serially, after the `jvm` container goal.
+
+### Dynamic destination
+
+You'll notice that the destination of the resulting Docker image is a static location, which is less than ideal, because it means that you can't (easily) store multiple versions of the images. To alleviate this, container goals also have a `callback` property, which are executed right before the container goal. You can use this function to fetch information about the project that the SDM is operating within:
+
+```typescript
+  return {
+    jvm: { ... },
+    docker: {
+      goals: [
+        callback: async (reg, proj) => {
+            // remove non-alphanumeric characters
+            const safeOwner = proj.id.owner.replace(/[^a-z0-9]+/g, "");
+            const dest = `${safeOwner}/${proj.id.repo}:${proj.id.sha}`;
+            reg.containers[0].args.push(`--destination=${dest}`);
+            return reg;
+        },
+        container("kaniko", {
+          input: ["target-jar"],
+          containers: [{
+            name: "kaniko",
+            image: "gcr.io/kaniko-project/executor:v0.10.0",
+            args: [
+                "--dockerfile=Dockerfile",
+                "--context=dir://atm/home",
+                "--no-push",
+                "--single-snapshot"
+            ],
+          }],
+        }),
+      ],
+      test: ...,
+      dependsOn: ...
+    }
+  };
+```
+
+`reg`refers to the container goal registrations which are defined as the second argument to the`container`function, while`proj`is the project the SDM is running against. From the`proj`, it's possible to get the name of the underlying repository, which we can use to define a more unique resulting image, based on the SHA of the last pushed commit.
+
+## Complete sample
```

Comment:
Maybe this should point to a sample in the samples repo so it is more likely to stay working?